### PR TITLE
Adding Flatpak module

### PR DESCRIPTION
**Pull request by JayKayy**
##### SUMMARY
This change adds the `flatpak` module. This initial piece is a simple install/uninstall module for flatpaks and flatpak remotes.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
flatpak

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.4.0 (devel e3698d8d37) last updated 2017/07/26 15:26:31 (GMT -400)
  config file = /home/jkwiatko/.ansible.cfg
  configured module search path = [u'/home/jkwiatko/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jkwiatko/github/ansible/lib/ansible
  executable location = /home/jkwiatko/github/ansible/bin/ansible
  python version = 2.7.13 (default, Jun 26 2017, 10:20:05) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```
##### ADDITIONAL INFORMATION
example usage:
```
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://jgrulich.fedorapeople.org/telegram/telegram.flatpakref state=present'
localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.telegram.desktop/x86_64/master    system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://git.gnome.org/browse/gnome-apps-nightly/plain/gedit.flatpakref state=present'
localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.gnome.gedit/x86_64/master         system,current
org.telegram.desktop/x86_64/master    system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://git.gnome.org/browse/gnome-apps-nightly/plain/gedit.flatpakref state=absent'localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.telegram.desktop/x86_64/master    system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://jgrulich.fedorapeople.org/telegram/telegram.flatpakref state=absent'
localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current

jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.gnome.gedit/x86_64/master         system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=org.gnome.gedit state=absent'localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
[jkwiatko ansible]$ 

```

### CODE REVIEW

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (sysadmind) [inline lib/ansible/modules/packaging/os/flatpak.py]: I believe that this section should start on the same line as the dash instead of the line beneath from looking at other modules.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (sysadmind) [inline lib/ansible/modules/packaging/os/flatpak.py]: This should also start on the line above if the other is moved.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (sysadmind) [inline lib/ansible/modules/packaging/os/flatpak.py:16]: The flatpak command is listed as a requirement but the module doesn't check if it is before proceeding. It would be a good idea to `module.fail_json()` with a useful message when that occurs. You can actually even get the full path with the function `module.get_bin_path()` as illustrated here: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/system/service.py#L451
That would also get you a full path to execute so that you don't run into relative path issues.

**Review by sysadmind — COMMENTED** [review]
Overall looks pretty good. Just a few doc recommendations to follow other modules and some useful feedback to users on errors.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (sysadmind) [inline lib/ansible/modules/packaging/os/flatpak.py]: Each of these functions creates a command but uses the relative executable `flatpak` but you have the absolute path below as `location` and you fail if it's not found. You could pass that location as an agrument or create a class wrapper around the functions. It would ensure that the initial flatpak path would always be used and possibly prevent environment issues.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (andytom) [inline lib/ansible/modules/packaging/os/flatpak.py]: Should this be ```... for state=`present` ``` ?

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (andytom) [inline lib/ansible/modules/packaging/os/flatpak.py]: As above should this be ```...reverse DNS state=`absent` ...```?

**Review by andytom — COMMENTED** [review]
Few minor changes to the docs.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py]: You are correct on both of these accounts

**Comment by sysadmind** [comment]
shipit

@JayKayy I'm really happy to see flatpak support land in Ansible. Nice work.

**Comment by JayKayy** [comment]
Thanks. I certainly found it useful for my personal laptop setup :)

**Review by andytom — APPROVED** [review]
shipit

**Comment by pmakowski** [comment]
shipit

**Comment by jhawkesworth** [comment]
bot_status

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: This doesn't need to be quoted.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: This doesn't need to be quoted.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: This doesn't need to be quoted.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: This doesn't need to be quoted.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: Convert the task comments into the task names:

```
- name: Install the spotify flaptalk
...
```

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: There should be no empty line here.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: There should be new line behind this line.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: There should be 2 new lines behind this line.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: Either remove this empty line OR add the same empty line into all function in front of the return command (preferred solution).

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: You don't have to quote it even if the text spans multiple lines. Also, it should use the standard formatting for names and values of options and modules as described [here](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#formatting-options).

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: The same like above.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: No quotes here.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: No quotes here.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: Capital letter in the value.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: No quotes here.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: No quotes here.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: No quotes here.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: This should be wrapped like this:

```
            state=dict(
                required=False, default="present", choices=['present', 'absent'])
```

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: No backslashes here.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: There should be a new line above this line.

**Review by jtyr — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: You can delete these empty lines.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: You can remove the backticks when using the other decoration.

**Comment by JayKayy** [comment]
I have messed up my branch running into merge conflicts with files i didnt modify. I will be fixing these shortly.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:72]: This should be indented by 2 spaces. The same bellow.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:25]: You can drop the backticks around the `http(s)`.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:7]: I would format this like this:

```
ANSIBLE_METADATA = {
    'metadata_version': '1.0',
    'status': ['preview'],
    'supported_by': 'community'
}
```

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:96]: What about to insert new line in front of the `if`? It would make nice visual separation between the variable definitions and the condition. The same at any place bellow.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:18]: This should be indented only by 2 spaces.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:21]: This should be indented only by 2 spaces.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:46]: I would insert one new line in front of this.

**Review by jtyr — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [inline lib/ansible/modules/packaging/os/flatpak.py:136]: Done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:96]: This makes sense to me.

Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:96]: Done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:72]: done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:46]: done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:25]: done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:21]: done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:18]: done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (JayKayy) [reply lib/ansible/modules/packaging/os/flatpak.py:7]: done

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: I would put the final cruelly bracket on a new line:

```
    ...
    'supported_by': 'community'
}
```

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py:218]: What about to separate the `if` blocks here as well as you did above? The same bellow.

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (jtyr) [inline lib/ansible/modules/packaging/os/flatpak.py]: I kind of don't like returning from a function from a condition. What about to define `code` variable which you assign in the condition and then return at the end of the function?

```
    ...
    output = flatpak_command(command)
    code = 0

    if 'error' in output and 'already installed' not in output:
        code = 1

    return code, output
```

The same anywhere bellow.

**Review by jtyr — CHANGES_REQUESTED** [review]

**Comment by JayKayy** [comment]
Great feedback and thank you for the formatting/syntactical help. Any more feedback?

**Review by sysadmind — APPROVED** [review]
Thanks for all of the updates!

**Comment by JayKayy** [comment]
Are there any more TODO's on this?

**Comment by jctanner** [comment]
@JayKayy you still have a merge commit that can only be resolved by following the rebase instructions posted by the bot previously.

https://github.com/ansible/ansible/commit/55e8c16ff71107f64c791f5960b228a00204cf16 --> "2 parents 52e8d92 + e419a55"

Until you resolve that, we will not merge this PR.

**Comment by mattclay** [comment]
This PR contains many changes unrelated to the module being added.

needs_rebase

File: lib/ansible/modules/packaging/os/flatpak.py
Comment (dagwieers) [reply lib/ansible/modules/packaging/os/flatpak.py:21]: Actually this shouldn't be indented at all. Lists can start from the same indentation as their parent.
And the general rule is, don't add things that aren't needed (so don't quote strings if they're not needed, or don't indent if it's not needed).

**Comment by meebey** [comment]
@JayKayy will you rebase and revive this PR later? Easiest would be to just add the module file on a fresh topic branch as I see there are some merge conflict markers committed

Never mind, just found the other PR where you continue: ansible/pull/28742

**Pull request closed without merging**
